pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,186 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	podName             = "name"
+	podNameSpace        = "namespace"
+	ownerSelfLink       = "owner"
+	podHostName         = "node"
+	podProvisioner      = "provisioner"
+	podHostZone         = "zone"
+	podHostArchitecture = "arch"
+	podHostCapacityType = "capacity_type"
+	podHostInstanceType = "instance_type"
+	podPhase            = "phase"
+)
+
+var (
+	podGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "pods",
+			Name:      "state",
+			Help:      "Pod state.",
+		},
+		labelNames(),
+	)
+)
+
+// Controller for the resource
+type Controller struct {
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName]prometheus.Labels
+}
+
+func init() {
+	crmetrics.Registry.MustRegister(podGaugeVec)
+}
+
+func labelNames() []string {
+	return []string{
+		podName,
+		podNameSpace,
+		ownerSelfLink,
+		podHostName,
+		podProvisioner,
+		podHostZone,
+		podHostArchitecture,
+		podHostCapacityType,
+		podHostInstanceType,
+		podPhase,
+	}
+
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName]prometheus.Labels),
```

Comment:
Why is naming styled differently on this one?

Reply:
The pod controller only keeps a map of labels, however, the node controller keeps a map of slice of labels. I don't want to call the one in node controller just labels since that will confuse the layer of structure here.